File: LeetCodeProblemsCpp/LeetCodeProblemsCpp/Problems/LeetCode2197ReplaceNonCoprimeNumbersInArray.cpp
#include "LeetCode2197ReplaceNonCoprimeNumbersInArray.h"

#include <algorithm>
#include <stack>

#include "Auxilaries.h"
using namespace std;
int GCD(int a, int b)
{
    if (b == 0)
        return a;
    return GCD(b, a % b);
}
int LCM(int a,int b)
{
    long result=a;
    result*=b;
    result/=GCD(a,b);
    return result;
}
vector<int> LeetCode2197ReplaceNonCoprimeNumbersInArray::replaceNonCoprimes(vector<int>& nums)
{
    stack<int> numsStack{};
    numsStack.push(nums[0]);
    for(int i=1,e= static_cast<int>(nums.size());i<e;i++)
    {
        auto num = nums[i];
        auto top = numsStack.top();
        while (GCD(top,num)>1 )
        {
            num=LCM(top,num);
            numsStack.pop();
            if (numsStack.empty())
                break;
            top = numsStack.top();
        }
        numsStack.push(num);
    }
    vector<int> result{};
    result.reserve(numsStack.size());
    while (!numsStack.empty() )
    {
        result.push_back(numsStack.top());
        numsStack.pop();
    }
    ranges::reverse(result);
    return result;
}

void LeetCode2197ReplaceNonCoprimeNumbersInArray::Run()
{
    vector<int> nums{6,4,3,2,7,6,2};
    Auxilaries::PrintVector(replaceNonCoprimes(nums));
}

File: LeetCodeProblemsGo/LeetCode2197ReplaceNonCoprimeNumbersInArray/LeetCode2197ReplaceNonCoprimeNumbersInArray.go
package LeetCode2197ReplaceNonCoprimeNumbersInArray

func gcd(a, b int) int {
	if b == 0 {
		return a
	}
	return gcd(b, a%b)
}
func lcm(a, b int) int {
	return a * b / gcd(a, b)
}
func replaceNonCoprimes(nums []int) []int {
	numsStack := []int{}
	top, current := 0, 0
	for _, num := range nums {
		current = num
		if len(numsStack) > 0 {
			top = numsStack[len(numsStack)-1]
			for gcd(top, current) > 1 {
				current = lcm(top, current)
				numsStack = numsStack[:len(numsStack)-1]
				if len(numsStack) == 0 {
					break
				}
				top = numsStack[len(numsStack)-1]
			}
		}
		numsStack = append(numsStack, current)
	}
	return numsStack
}
